File: 0198-house-robber/0198-house-robber.py
class Solution:
    def house(self,nums,idx,dp): 
        if idx>=len(nums): 
            return 0 
        if dp[idx]!=-1: 
            return dp[idx]
        # each rob and i+2 or don't and i+1 
        a = nums[idx]+self.house(nums,idx+2,dp) 
        b = self.house(nums,idx+1,dp)
        dp[idx] = max(a,b)
        return dp[idx]

    def rob(self, nums: List[int]) -> int:
        n = len(nums) 
        dp=[-1]*n 
        res = 0
        next1,next2=None,None 
        for i in range(n-1,-1,-1):
            take,notake = nums[i],0
            if i+2<n: 
                take+=next2
            if i+1<n:
                notake+=next1
            curr=max(take,notake)
            res=max(res,curr)
            next2=next1 
            next1=curr
        return res
        # return self.house(nums,0,dp)

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int houserob(vector<int>& nums,int idx){
        if(idx<0) return 0;
        int a = nums[idx]+houserob(nums,idx-2);
        int b = houserob(nums,idx-1);
        return max(a,b);
    }

    int rob(vector<int>& nums) {
        if(nums.size()<2) return nums.back();
        int n = nums.size();
        vector<int> dp(n,0);
        dp[0]=nums[0];
        dp[1]=max(nums[0],nums[1]);
        for(int i=2;i<nums.size();i++){
            dp[i]=max(nums[i]+dp[i-2],dp[i-1]);
        }
        return dp.back();
        // return houserob(nums,n-1);
    }
};